rfcs/20190225-tf-on-demand.md

```diff
@@ -0,0 +1,191 @@
+# TensorFlow On Demand
+
+| Status        | Proposed      |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Gunhan Gulsoy (gunan@google.com), Hye Soo Yang (hyey@google.com) |
+| **Sponsor**   | Gunhan Gulsoy (gunan@google.com)                |
+| **Collaborators** | Christoph Goern (goern@redhat.com), Subin Modeel (smodeel@redhat.com)         |
+| **Updated**   | 2019-02-25                                           |
+
+## Objective
+
+This document proposes a system to build optimized binaries in the cloud and deliver the artifacts to users of TensorFlow. With this system, we aim to achieve the following:
+
+*   Hide the complexity of the build system of TF from beginner and intermediate users.
+*   Improve the out-of-the-box performance for users.
+*   Build on the strength of TensorFlow's community partners.
+
+## Motivation
+
+### Overview
+
+TF build is difficult and the friction present during building phase is apparent in our user experiences. For a successful build, users needs to be aware of requirements and configurations that they might not be too familiar with. This can be extremely challenging for beginner and intermediate users.
+
+Currently, TF python release artifacts are hosted on PyPI as a single source of TF binaries for download. However, PyPI is quite limited in capability; it is unable to recognize and/or deploy on user machines.
+
+Historically, this has been a problem when trying to satisfy the following requirements for our artifacts:
+
+*   **Portability**: The artifacts should be able to run on as many platforms as possible.
+*   **Performance**: The artifacts should run as fast as possible.
+*   **Size**: The artifacts should be as small as possible in size.
+
+To achieve the above, we propose a system that takes inputs from the user, build artifacts based on the inputs and send the output back to the user. We are proposing two user endpoints to the system:
+
+1.  Command line tool (python)
+1.  Web interface
+
+An overview of the system:
+
+![drawing](https://docs.google.com/drawings/d/11jAVBtR4nV4bkDVW1WrHhLXdzX2ZouOPwsztW2hZQz4/export/png)
+
+A system flow chart summarizing the ordering of events for all use cases of the system:
+
+![drawing](https://docs.google.com/drawings/d/1lSSHaYktst8MNTqF860cPlkfbgc3Ft_t_8PWzH0VkW4/export/png)
```

Comment:
The diagrams seems of low resolution...

Reply:
All diagrams are updated with better resolution!